src/rust/storage/seg/src/hashtable/mod.rs

```diff
@@ -109,6 +109,131 @@ pub(crate) struct HashTable {
     next_to_chain: u64,
 }
 
+struct Iter<'a> {
+    hashtable: &'a HashTable,
+    bucket_id: usize,
+    item_slot: usize,
+    chain_len: usize,
+    chain_idx: usize,
+    finished: bool,
+}
+
+struct IterMut<'a> {
+    ptr: *mut HashBucket,
+    bucket_id: usize,
+    item_slot: usize,
+    chain_len: usize,
+    chain_idx: usize,
+    finished: bool,
+    _marker: std::marker::PhantomData<&'a mut u64>,
+}
+
+impl<'a> Iter<'a> {
+    fn new(hashtable: &'a HashTable, hash: u64) -> Self {
+        let bucket_id = (hash & hashtable.mask) as usize;
+        let bucket = hashtable.data[bucket_id];
+        let chain_len = chain_len(bucket.data[0]) as usize;
+
+        Self {
+            hashtable,
+            bucket_id,
+            item_slot: 1,
```

Comment:
yeah I came to a similar conclusion that we should just start on the 1st item as the 0th one is the bucket info, right?